File: JAVA/DataStructures/Arrays/MaximumSubarray.java
public class MaximumSubarray {
    // Question Link:
    // https://leetcode.com/problems/maximum-subarray
    public static void main(String[] args) {
        int[] arr = {-2,1,-3,4,-1,2,1,-5,4};
        System.out.print(maxSubArray(arr));
    }

    static int maxSubArray(int[] nums) {
        // Using Kadane's Algorithm:
        int sum = 0;
        int max = nums[0];
        for(int i=0; i < nums.length; i++) {
            sum += nums[i];
            if(sum > max) {
                max = sum;
            }
            if(sum < 0) {
                sum = 0;
            }
        }
        return max;
    }

}

File: PYTHON/ProblemsAndSolutions/maximumSubarray.py
# Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

# A subarray is a contiguous part of an array.

 

# Example 1:

# Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
# Output: 6
# Explanation: [4,-1,2,1] has the largest sum = 6.

def maxSubArray(nums):
    maxSum = sum(nums)
    curr = 0
    for i in range(len(nums)):
        curr += nums[i]
        maxSum = max(maxSum, curr)
        if curr < 0:
            curr = 0
    return maxSum

arr = list(map(int, input().split()))
print("Maximum Subarray is : ", end= " ")
print(maxSubArray(arr))